File: Java/flesch.java
//Colton Steinbeck//
//Flesh Readibilty In Java//
import java.lang.*;
import java.util.*;
import java.io.*;

public class flesch
{    

    public static void main(String[] args) throws IOException
    {
        //Takes the users command line input
    System.out.println("Please enter specific file locations for translations: ");
    Scanner tScan = new Scanner(System.in);
    String translations = tScan.nextLine();

    //Takes the file path puts these words into the array//
    Scanner scan = new Scanner(new File(translations));
     
    ArrayList<String> vals= new ArrayList<String>();
    
        while(scan.hasNext())
        {
            String words= scan.next();
            vals.add(words);
        
        }    
        //Pulls the Dale-Chall words into storage 
    Scanner dc = new Scanner(new File("/pub/pounds/CSC330/dalechall/wordlist1995.txt"));
    Map<String, Object> dcStorage = new HashMap<String,Object>();
        
        while(dc.hasNextLine()) 
        {
            String word = dc.nextLine();
            dcStorage.put(word, null);
        }

        //Instanties Variables
        int words = 0;
        int syllables = 0;
        int sentences = 0;
        int diffWord = 0;
 

        //Goes through the whole array of the Bible (Main Array)
        
        for(int i=0; i<=vals.size()-1; i++)
        {
            String token = vals.get(i);
            token = token.toLowerCase();
        try
        { //Try-Catch loop so there is no OutOfBoundsException for the ArrayList
             
            //Check for punctuation
           for(int j=0; j<=token.length();j++)
           {
               char punc = token.charAt(j);
               if(punc == '.' || punc =='!' || punc == '?' || punc == ':' || punc == ';' )
               {
                    sentences++;
               }
           }
        }
        catch(StringIndexOutOfBoundsException e)
        {
    
        } 
        if (!isNumber(token))
        {
           words++        
           syllables = syllables + syllableCount(token); 
            if(!dcStorage.containsKey(token))
            {
                diffWord++; 
            }
        }
        }
    //Numeric calculations
    System.out.println("Word count: " + words);
    System.out.println("Sentence count: " + sentences);
    System.out.println("Syllable count: " + syllables);
    System.out.println("Difficult Word count: " + diffWord);

    double alpha = ((double)syllables / (double)words);
    double beta = ((double)words / (double)sentences);
    double gamma = ((double)diffWord / (double)words);

    double flesch = 206.835 - (alpha * 84.6) - (beta * 1.015);
    double grade = (alpha * 11.8) + (beta * 0.39) - 15.59;
    double readability;
    if (gamma > 0.05)
    {
        readability = ((gamma * 100.0) * 0.1579) + (beta * 0.0496) + 3.6365;
    }
    else
    {
        readability = ((gamma * 100.0) * 0.1579) + (beta * 0.0496);
    }
    System.out.println("The Flesch Readability index is: " + flesch);
    System.out.println("The Flesh-Kincaid Grade Level index is: " + grade);
    System.out.println("The Dale-Chall Readability Score is: " + readability);
}
    //Checks if it is a vowel
    private static boolean isVowel(char letter)
    {
        return letter == 'a' || letter == 'e' || 
               letter == 'i' || letter == 'o' ||
               letter == 'u' || letter == 'y';
    }

    //Counts the number of syllables in the 
    public static int syllableCount(String token)
    {
    int syllableCount=0;
    boolean found=false;
    //For each index until the length of the strin
    for(int i = 0; i < token.length(); i++)
    
    {
    
    char letter=token.charAt(i);
    letter=Character.toLowerCase(letter);
    
    
        //Checks if the letter is a vowel
        if(isVowel(letter)==true)
        {
        
        found=true;
        syllableCount++;
        
            if(i+1<token.length()) //if the next letter doesnt exceed the length of the string
            
            {
                char nextLetter=token.charAt(i+1); //Increments to the next letter
                nextLetter=Character.toLowerCase(nextLetter); //lowercase
                if(isVowel(nextLetter)==true) //if vowel is next letter
                {
                    syllableCount--;
                
                }
                
            }
        
        }
        //2nd Case: If the letter 'e' is the last character
        if( (letter== 'e') && (i+1>token.length()))
        {
            syllableCount--;
        }
    
    }
    
        if(found==false)
        {
            syllableCount++;
        }
            
    found=false;
    return syllableCount;
    }
    
    //Checks if the string passed is a integer
    private static boolean isNumber(String value)
    {
        try 
        {
        Integer.parseInt(value);
        
        return true;    
        } 

        catch (Exception e) 
        {
            return false;
        }
    }

 
}//End of Class

File: C++/flesch.cpp
//Colton Steinbeck//
//Flesh Readibilty In C++//
#include <iostream>
#include <fstream> //Reading the 
#include <cstring>
#include <unordered_map> //Used for Dale-Chall
#include <vector>
#include <cmath>
using namespace std;

//Checks if the character is a vowel
bool IsVowel (char letter)
{
if(letter == 'a' || letter == 'e' || letter == 'i' || letter == 'o' || letter == 'u' || letter =='y')
{
    return true;
}
return false;
} 



//String Cleanup 
string stringCleanup(string word)
{
    string tmp = "";
    for(int i=0; i<word.size();i++)
    {
        
        tolower(word[i]);
        if(isalpha(word[i]))
        {
            tmp+=(word[i]);
        }
    }
    return tmp;
}

//Method to count the syllables in each word
int syllableCount( string token ) 
{
int syllable = 0;
    bool found=false;
        
    for(int i = 0; i < token.length(); i++)
    
    {
    //Breaks down word to the char letter
    char letter=token[i];
    tolower(letter); //lowercases it
        if(IsVowel(letter)==true) //Checks using IsVowel method
        {
        found=true;
        syllable++;
            if(i+1<token.length()) //As long as the next letter doesn't exceed the length of the word
            {
                char nextLetter=token[i+1];
                nextLetter=tolower(nextLetter);
                if(IsVowel(nextLetter)==true) //As long as the current letter is a vowel and the next letter is not a vowel
                {
                    syllable--;
                
                }  
            }
        }
        //2nd Case- Deals with 'e' at the end of the word//
        if( (letter== 'e') && (i+1>token.size()))
        {
            syllable--;
        }
    
    }
            //3rd Case- If it claims there is no syllable
            if(found==false)
            {
                syllable++;
            }
            
            found=false;
            return syllable;
}

//Checks if the value is an integer
bool isNum(string token)
{
    int count=0;
for(int i=0;i<token.size();i++)
{
    if(isdigit(token[i])) //If char is a digit
    {
        count++;
        
    }
}
    if(count==token.size()) //Last Char check
    {
        return true;
    }
    return false;
}

void debug(int n) //Debugging Method
{
    cout << "we made it to " << n << "!" << endl;
}

//Main Function/////
int main()
{

    //Command to input file location
    cout<<"Please enter specific file locations for translations:  "<<endl;
    string input;
    cin>>input;

    //Checks if the file has opened
    ifstream BibleFile(input);
        //Error Handling//
        if (BibleFile.is_open())
        {
		    cout << "The Bible Translation opened correctly." << endl;
        }
        if(BibleFile.fail())
        {
            cout<<"File cannot be accessed, please try again"<<endl;
        }
            //Imports the Dale-Chall List of Difficult Words
    ifstream dalechall("/pub/pounds/CSC330/dalechall/wordlist1995.txt");
    	if (dalechall.is_open())
        {
		    cout << "The Dale-Chall list opened correctly." << endl;
        }
	    else 
        {
            cout << "Unable to open Dale-Chall list of words." << endl; 
        }


    //Creation of Vectors for the Translations
    vector<string> vals;
    string tmp;
    while(BibleFile >> tmp) 
    {
        vals.push_back(tmp); //Adds vals of the translation to the vector
    }
    unordered_map <string,int> dcStorage; //Storage for the Dale-Chall List
    string temp;
    int iteration=0;
    while(dalechall >> temp)
    {
        iteration++;
        for(int i=0;i<temp.size();i++)  
        {
            if(ispunct(temp[i]) && isalpha(temp[i])) //Punctuation and Alpha-Numeric Check
            {
                temp.erase(i--,1);//Removes the value if happens to be puncuation
            }
            temp[i]=tolower(temp[i]); //lowercases the words in Dale-
        } 
        dcStorage.insert(make_pair(temp, iteration)); //Inserts Map Pairing for value and placement
        
    }

    int word=0;
    int syllable=0;
    int sentence=0;
    int diffWord=0;

    //Looping through the vector for the translation
    for(int i=0; i<=vals.size()-1; i++)
    {
            string token = vals[i]; //Breaks down word by word
        try
        { //Try-Catch to make sure there is no OutOfBounds Exceptions
             
            //Sentence Check
           for(int j=0; j<=token.length();j++)
           {
               tolower(token[j]); //Lowercase each letter
               char punc = token[j];
               if(punc == '.' || punc =='!' || punc == '?' || punc == ':' || punc == ';' )
               {
                    sentence++;
               }
            }
        }
        catch(int e)
        {
    
        } 
        if (!isNum(token)) //Check if the word is a number
        {
         
            word++;
        
            syllable = syllable + syllableCount(token); //Checks using the syllable method
            token= stringCleanup(token); //Using method to help make more accurate of token
            if(dcStorage.find(token) == dcStorage.end()) //Checks the word to the DC List
            {
                diffWord++;
                
            }
        }
    }

    //Numeric Calculations
	cout << "Word count: " << word << endl;	
	cout << "Sentence count: " << sentence << endl;
	cout << "Syllable count: " << syllable << endl;
	cout << "Difficult Word count: " << diffWord << endl;

	double alpha = ((double)syllable / (double)word);
	double beta = ((double)word / (double)sentence);
	double gamma = ((double)diffWord / (double)word) * 100;

	double index = (206.835 - (alpha * 84.6) - (beta * 1.015));
	double grade = ((alpha * 11.8) + (beta * 0.39) - 15.59);

	cout << "The Flesch index score is: " << round(index) << endl;
	cout << "The Flesch Kincaid Grade Level index is: " << grade << endl;

	double readability;

	if (gamma > 5.0)
		readability = (gamma * 0.1579) + (beta * 0.0496) + 3.6365;
	else
		readability = (gamma * 0.1579) + (beta * 0.0496);

	cout << "The Dale-Chall Readability score is: " << readability << endl; 

    //Closes the file after it is used
	BibleFile.close();	

	return 0;
}















